### PR TITLE
Allow to use "\" in the file of parameters

### PR DESCRIPTION
**Pull request by asartori86**
With this PR it is possible to split the definition of a parameter entry on multiple lines in the .prm file.

see the file `tests/bits/prm/parameter_handler_2_backslash.prm` at the bottom

### CODE REVIEW

File: source/base/parameter_handler.cc [hunk 1]
```diff
@@ -1382,10 +1382,44 @@ bool ParameterHandler::read_input (const std::string &filename,
...
+      bool is_concatenated(false);
+      while (std::getline(file, str))
+        {
+          // if this str must be concatenated with the previous one
+          // we strip the whitespaces
+          if (is_concatenated)
+            {
+              const std::size_t str_begin = str.find_first_not_of(" \t");
+              if (str_begin == std::string::npos)
+                return ""; // no content
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: I don't think you want to exit the entire function just because you've found an empty continuation line :-)

**Comment by asartori86** [comment]
Oops!
 now should be fixed. Thanks

**Comment by bangerth** [comment]
@asartori86 -- nice, thanks. This has been a long-standing enhancement to add, see #333 .

I have a few comments:
1/ In your current code, you track whether a line is a continuation line of the previous one, but the only thing you do if that is the case is trim whitespace from the front and back of the continuation line. I suppose you do this trimming because you want to make it look like any leading space on the continuation line isn't there. (Or is this true?) If so, you need to only trim spaces at the *front* of the continuation line, but you *also* have to trim spaces at the *end* of the continued line.

2/ Your code currently only checks whether there is a backslash *anywhere*, replaces it by a space, and then considers the line continued on the next line. You want to be more specific and only consider backslashes *at the end of the line* as special.

3/ I think the code is in the wrong place since it only handles the situation where the input comes from a file. My suggestion would be to keep the code that reads the file into a string, but then move the code that processes the continuation line into the `read_input_from_string()` function so that one also gets the continuation line processing when calling that function. Your new code could simply be some kind of pre-processing within that function.

**Comment by QiaoLei-88** [comment]
1. I think it would be better to add some off-design cases in the test, such as empty line, empty line with a backslash, a backslash not at the end, multiple backslashes, two concatenated parameters separated by a not concatenated parameter.

2. I suggest not to replace the backslash by a space, just remove it. If user want to have space, they can put a space before the backslash. And spaces before backslash should not be trimmed because they are not logical end spaces.

**Comment by bangerth** [comment]
On 09/03/2015 08:26 AM, Lei Qiao wrote:
> I think it would be better to add some off-design cases in the test, such as
> empty line, empty line with a backslash, a backslash not at the end, multiple
> backslashes, two concatenated parameters separated by a not concatenated
> parameter.

Or even better, have multiple testcases that each test one particular aspect 
of it all. There is no such thing as having too many testcases.

**Comment by asartori86** [comment]
> 3/ I think the code is in the wrong place since it only handles the situation where the input comes from a file. My suggestion would be to keep the code that reads the file into a string, but then move the code that processes the continuation line into the read_input_from_string() function so that one also gets the continuation line processing when calling that function. Your new code could simply be some kind of pre-processing within that function.

@bangerth, Thanks for your commments. could you check if I correctly understood what you suggested? thanks

File: source/base/parameter_handler.cc [hunk 2]
```diff
@@ -1382,10 +1382,9 @@ bool ParameterHandler::read_input (const std::string &filename,
   try
     {
       std::string openname = search.find(filename);
-      std::ifstream input (openname.c_str());
-      AssertThrow(input, ExcIO());
-
-      return read_input (input, filename);
+      std::ifstream file (openname.c_str());
+      AssertThrow(file, ExcIO());
+      return read_input_from_string (openname.c_str());
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: I don't think this would work. Did you test this? You pass the *file name*, not the *content of the file* to `read_input_from_string`. My idea was that you simply read the entire file into a string (basically like you did in the first round of review, but without the processing) and then pass that string to `read_input_from_string()`.

File: source/base/parameter_handler.cc [hunk 3]
```diff
@@ -1409,7 +1408,41 @@ bool ParameterHandler::read_input_from_string (const char *s)
 {
   // create an istringstream representation and pass it off
   // to the other functions doing this work
-  std::istringstream in (s);
+
+  // concatenate the lines ending with "\"
+
+  std::ifstream file(s);
+  std::string str;
+  std::string file_contents;
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: This changes the meaning of the function. It used to just get the *contents* of a file as a string, but now it gets a *file name*.

File: source/base/parameter_handler.cc [hunk 4]
```diff
@@ -1409,7 +1408,41 @@ bool ParameterHandler::read_input_from_string (const char *s)
...
+      if (std::find(str.begin(),str.end(),'\\') == str.end())
+        {
+          is_concatenated = false; // next line must be concatenated = false
+          file_contents += str;
+          file_contents.push_back('\n');
+        }
+      else
+        {
+          str.erase(str.find_last_not_of(" \t") + 1); // trim withespaces after "\"
+          str.erase(str.find_last_of('\\')); // remove "\" from the end of line
```
Comment (bangerth) [inline source/base/parameter_handler.cc]: You somehow need to test that the backslash really is the last character on the line. Right now, you're simply deleting the last occurrence, but that might be somewhere in the middle where it is desired.

I think you need to check in the `if` condition on line 1431 that indeed the backslash is the last non-space character on the line, not whether there *is* or *is not* a backslash on the line.

File: tests/bits/prm/parameter_handler_backslash_04.prm [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#Listing of Parameters
+#---------------------
+subsection Testing
+  set Function = a,\ b,
+                 c
```
Comment (bangerth) [inline tests/bits/prm/parameter_handler_backslash_04.prm]: this should actually yield an error, but it doesn't currently because you don't check *where* the backslash is -- as long as there *is* a backslash somewhere in the line, you treat it as a continued line, but that seems wrong to me.

**Comment by asartori86** [comment]
@bangerth, does it look better now? All the `parameter_handler*` tests passed on my laptop

File: tests/bits/parameter_handler_backslash_02.output [hunk 6]
```diff
@@ -0,0 +1,2 @@
+
+DEAL::a, b, c
```
Comment (QiaoLei-88) [inline tests/bits/parameter_handler_backslash_02.output]: I think there should be no space in between parameters.

File: source/base/parameter_handler.cc [hunk 7]
```diff
@@ -1409,7 +1418,40 @@ bool ParameterHandler::read_input_from_string (const char *s)
...
+    {
+      // if this str must be concatenated with the previous one
+      // we strip the whitespaces
+      if (is_concatenated)
+        {
+          const std::size_t str_begin = str.find_first_not_of(" \t");
+          if (str_begin == std::string::npos)
+            str = ""; // no content
+          else
+            str.erase(0,str_begin); // trim whitespaces from left
```
Comment (QiaoLei-88) [inline source/base/parameter_handler.cc]: ```str.erase(0,str_begin+1)``` should remove the extra space.

Comment (asartori86) [inline source/base/parameter_handler.cc]: Thanks for pointing it out. The file `parameter_handler_backslash_02.output` was the old version. In the new one there are no spaces. Yet the test was successful

Comment (QiaoLei-88) [inline source/base/parameter_handler.cc]: Ah,  my bad.

File: tests/bits/parameter_handler_backslash_03.expect=diff.output [hunk 8]
```diff
@@ -0,0 +1,2 @@
+
+DEAL::a, b, c
```
Comment (bangerth) [inline tests/bits/parameter_handler_backslash_03.expect=diff.output]: why is the file named this way? why not just `parameter_handler_backslash_03.output`?

File: tests/bits/parameter_handler_backslash_04.expect=diff.output [hunk 9]
```diff
@@ -0,0 +1,2 @@
+
+DEAL::a, b, c
```
Comment (bangerth) [inline tests/bits/parameter_handler_backslash_04.expect=diff.output]: same here

File: tests/bits/parameter_handler_backslash_03.expect=diff.output (hunk 8 above)
Comment (asartori86) [inline tests/bits/parameter_handler_backslash_03.expect=diff.output]: because the `\` is put in the wrong place in file `parameter_handler_backslash_03.prm` and the output will contain just the `a`. same for test 04

Comment (bangerth) [inline tests/bits/parameter_handler_backslash_03.expect=diff.output]: I see now -- I didn't realize that we support this syntax...

Comment (asartori86) [inline tests/bits/parameter_handler_backslash_03.expect=diff.output]: :)

**Comment by bangerth** [comment]
OK, looks good now. Can you (i) rebase to current master, (ii) place an entry into the "general improvements" section of `doc/news/changes.h`, (iii) squash everything into one commit?

This is ready to merge then.

**Comment by asartori86** [comment]
@bangerth, done :)

**Comment by bangerth** [comment]
Great, thanks! And welcome among the authors of deal.II!

**Comment by asartori86** [comment]
Thanks to you :)
